cluster/replication/shard_replication_engine.go

```diff
@@ -25,207 +27,732 @@ import (
+		err := e.producer.Produce(engineCtx, e.opsChan)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			e.logger.WithField("producer", e.producer).WithError(err).Error("stopping producer after failure")
+			producerErrChan <- err
+		}
+		e.logger.WithField("producer", e.producer).Info("replication engine producer stopped")
+	}, e.logger)
+
+	// Start one replication operations consumer.
+	e.wg.Add(1)
+	enterrors.GoWrapper(func() {
+		defer e.wg.Done()
+		e.logger.WithField("consumer", e.consumer).Info("starting replication engine consumer")
+		err := e.consumer.Consume(engineCtx, e.opsChan)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			e.logger.WithField("consumer", e.consumer).WithError(err).Error("stopping consumer after failure")
+			consumerErrChan <- err
+		}
+		e.logger.WithField("consumer", e.consumer).Info("replication engine consumer stopped")
+	}, e.logger)
+
+	// Coordinate replication engine execution with producer and consumer lifecycle.
+	var err error
+	select {
+	case <-ctx.Done():
+		e.logger.WithField("engine", e).Info("replication engine cancel request, shutting down")
+		err = ctx.Err()
+	case <-e.stopChan:
+		e.logger.WithField("engine", e).Info("replication engine stop request, shutting down")
+		// Graceful shutdown executed when stopping the replication engine
+	case producerErr := <-producerErrChan:
+		if !errors.Is(producerErr, context.Canceled) {
+			e.logger.WithField("engine", e).WithError(producerErr).Error("stopping replication engine producer after failure")
+			err = fmt.Errorf("replication engine producer failed with: %w", producerErr)
+		}
+	case consumerErr := <-consumerErrChan:
+		e.logger.WithField("engine", e).WithError(consumerErr).Error("stopping replication engine consumer after failure")
+		err = fmt.Errorf("replication engine consumer failed with: %w", consumerErr)
+	}
+
+	// Always cancel the replication engine context and wait for the producer and consumers to terminate to gracefully
+	// shut down the replication engine the both the producer and consumer.
+	engineCancel()
+	e.wg.Wait()
+	close(e.opsChan)
+	e.isRunning.Store(false)
+	return err
+}
+
+// Stop signals the replication engine to shut down gracefully.
+//
+// It safely transitions the engine's running state to false and closes the internal stop channel,
+// which unblocks the main loop in Start() and initiates the shutdown sequence.
+// Calling Stop multiple times is safe; only the first call has an effect.
+// Note that the ops channel is closed in the Start method after waiting for both the producer and consumers to
+// terminate.
+func (e *ShardReplicationEngine) Stop() {
+	if !e.isRunning.CompareAndSwap(true, false) {
+		return
+	}
+
+	// Closing the stop channel notifies both the producer and consumer to shut down gracefully coordinating with the
+	// replication engine.
+	close(e.stopChan)
+	if e.cancel != nil {
+		e.cancel()
+	}
```

Comment:
Follow-up to the comment above: are there cases when `e.cancel == nil`? Seems like both `e.stopChan` and `e.cancel` are always set in Start.